rmw_connext_shared_cpp/src/node.cpp

```diff
@@ -451,14 +486,10 @@ destroy_node(const char * implementation_identifier, rmw_node_t * node)
   }
 
   auto node_info = static_cast<ConnextNodeInfo *>(node->data);
-  if (!node_info) {
-    RMW_SET_ERROR_MSG("node info handle is null");
-    return RMW_RET_ERROR;
-  }
+  assert(node_info != nullptr);
   auto participant = static_cast<DDS::DomainParticipant *>(node_info->participant);
-  if (!participant) {
-    RMW_SET_ERROR_MSG("participant handle is null");
-  }
+  assert(participant != nullptr);
```

Comment:
why assert?

Reply:
Same as in https://github.com/ros2/rmw_fastrtps/pull/408#discussion_r453877085. There is no correct program that may pass a `rmw_node_t` structure that is not fully initialized. If `node_info` or `participant` are `nullptr`, the program is likely corrupt. One may argue though that the likelihood of memory corruption resulting in `nullptr` is low, and so the face value of `assert` vs. not doing anything is equally low. I'd still keep this boy.

Reply:
Which reminds me that we should document that after `rmw_destroy_node()`, using the pointer is UB.

Reply:
> One may argue though that the likelihood of memory corruption resulting in `nullptr` is low, and so the face value of `assert` vs. not doing anything is equally low. I'd still keep this boy.

Right, that's my general argument.  As a library, if it is something that we can reasonably check, then it is unlikely (though not impossible) that it is UB.  If it is UB, then there is nothing that we can do.  So to me, that argues for either doing nothing (there's nothing we can do about UB), or doing a nullptr check-and-return (which would be guarding against a caller of the API accidentally passing a nullptr in, nothing more).

Reply:
> As a library, if it is something that we can reasonably check, then it is unlikely (though not impossible) that it is UB.

I'd reverse that statement. If it is UB, then yes, it's unlikely we can guard ourselves against it. The fact that we can easily check for nullity here doesn't imply anything about it being UB or not. 

> So to me, that argues for either doing nothing (there's nothing we can do about UB), 

True in the most general case. I don't mind dropping the `assert` if you guys think it adds little value.

> doing a nullptr check-and-return (which would be guarding against a caller of the API accidentally passing a nullptr in, nothing more).

This scenario is not equivalent to an accidental `nullptr` argument (for which we do have a check, even though I have mixed feelings about dealing with logical errors in run-time). It can only happen if the `rmw_node_t` structure was intentionally or unintentionally corrupted. Guarding against such bogus code in run-time adds no value, and it only makes debugging harder when the already bad program eventually crashes further down the road.

Reply:
> There is no correct program that may pass a rmw_node_t structure that is not fully initialized. If node_info or participant are nullptr, the program is likely corrupt. One may argue though that the likelihood of memory corruption resulting in nullptr is low, and so the face value of assert vs. not doing anything is equally low. I'd still keep this boy.

My question is why not keeping the old error checking, I don't see an extra benefit of `assert` over normal error handling.

Reply:
Oof. We just had a rather long discussion with Chris (which I believe he'll post next). The main takeaway is, we shouldn't `assert` nor return an error code. We cannot guard against UB in general (so an `assert` against a particular UB case adds little value), and we shouldn't treat it as an expected or exceptional error condition, it's not (so no return code either, which IMHO is actually harmful in this case as it delays the crash of an already ill-formed or corrupt program).

Reply:
> Oof. We just had a rather long discussion with Chris (which I believe he'll post next)

I posted it over at https://github.com/ros2/rmw_fastrtps/pull/408#discussion_r454341097 , but I forgot we were having essentially the same conversation on two different PRs :).

Reply:
> The main takeaway is, we shouldn't assert nor return an error code. We cannot guard against UB in general

I'm ok with the change.
I understand that we cannot guard against general UB, e.g.: the user corrupting a structure with an invalid pointer.

If I'm going to dereference a pointer, I prefer checking that it's not NULL first in some way, but I don't mind strongly about it.

Reply:
> If I'm going to dereference a pointer, I prefer checking that it's not NULL first in some way, but I don't mind strongly about it.

Yeah, that's the usual reflex. And why I initially added an `assert`. But there's nothing special about a `NULL` pointer, besides the fact that it is guaranteed to always be invalid i.e. `ptr != NULL` does not imply `ptr` is valid. I'd assert at most, but otherwise do nothing.